File: data/projects/transit-multimodal/load.py
"""Load the `transit-multimodal` project network (Python).

Reads the CSVs in this folder and builds an undirected, weighted, multimodal
python-igraph object: neighborhoods are nodes; edges are transit links in two
modes (``metro`` and ``bus``). Because the same neighborhood pair can carry both
a metro and a bus link, the graph is a multiplex with parallel edges. The edge
weight is ``capacity`` (riders/hr). Run it straight (``python load.py``) for a
quick summary, or import ``load_transit()`` / ``load_lines()`` into your script.
"""
from __future__ import annotations

from pathlib import Path
import pandas as pd
import igraph as ig

HERE = Path(__file__).resolve().parent


def load_nodes() -> pd.DataFrame:
    """Node table: one row per neighborhood."""
    return pd.read_csv(HERE / "nodes.csv")


def load_edges() -> pd.DataFrame:
    """Edge table: one row per transit link (metro or bus)."""
    return pd.read_csv(HERE / "edges.csv")


def load_lines() -> pd.DataFrame:
    """Line/route lookup table (join onto edges by ``line``)."""
    return pd.read_csv(HERE / "lines.csv")


def load_transit(nodes: pd.DataFrame | None = None,
                 edges: pd.DataFrame | None = None) -> ig.Graph:
    """Build the undirected, weighted, multimodal graph (weight = ``capacity``).

    The same neighborhood pair may appear twice (a metro edge AND a bus edge) ->
    parallel edges. The ``mode`` edge attribute distinguishes them. To analyze a
    single mode, filter edges first; to collapse to one link per pair, use
    ``g.simplify(combine_edges={'weight': 'sum'})``.
    """
    if nodes is None:
        nodes = load_nodes()
    if edges is None:
        edges = load_edges()
    g = ig.Graph.DataFrame(edges, directed=False, vertices=nodes, use_vids=False)
    g.es["weight"] = g.es["capacity"]
    return g


if __name__ == "__main__":
    print("\n🚇 transit-multimodal (Python)")
    print("   Undirected multimodal transit; neighborhoods linked by metro & bus.\n")

    nodes = load_nodes()
    edges = load_edges()
    g = load_transit(nodes, edges)

    modes = edges["mode"].value_counts()
    print(f"✅ Loaded {len(nodes)} neighborhoods and {len(edges)} edges "
          f"({modes.get('metro', 0)} metro + {modes.get('bus', 0)} bus).")
    print(f"🔗 Directed: {g.is_directed()} | metro-served neighborhoods: "
          f"{int(nodes['has_metro'].sum())} | total seat capacity: "
          f"{edges['capacity'].sum():,} riders/hr")
    print("🎉 Graph ready. Object `g` is an undirected, weighted, multimodal igraph.")

File: data/projects/transit-multimodal/load.R
#' @name load.R
#' @title Load the `transit-multimodal` project network (R)
#' @description
#'
#' Reads the CSVs in this folder and builds an undirected, weighted, multimodal
#' igraph object: neighborhoods are nodes; edges are transit links in two modes
#' (`metro` and `bus`). The same neighborhood pair can carry both a metro and a
#' bus link, so the graph is a multiplex with parallel edges. The edge weight is
#' `capacity` (riders/hr). Run it straight (`Rscript load.R`) for a quick
#' summary, or `source()` it and call `load_transit()` in your own script.
#' `load_lines()` returns the line/route lookup table.

# 0. Setup ###################################################################
library(igraph)
library(here)

.dir <- function() here::here("data", "projects", "transit-multimodal")

#' Load the node table (one row per neighborhood).
load_nodes <- function() {
  read.csv(file.path(.dir(), "nodes.csv"), stringsAsFactors = FALSE)
}

#' Load the edge table (one row per transit link: metro or bus).
load_edges <- function() {
  read.csv(file.path(.dir(), "edges.csv"), stringsAsFactors = FALSE)
}

#' Load the line/route lookup table (join onto edges by `line`).
load_lines <- function() {
  read.csv(file.path(.dir(), "lines.csv"), stringsAsFactors = FALSE)
}

#' Build the undirected, weighted, multimodal transit graph.
#'
#' Edges are weighted by `capacity` (riders/hr). The `mode` edge attribute is
#' either `"metro"` or `"bus"`; the same neighborhood pair may appear as both
#' (parallel edges). Filter edges to analyze a single mode, or collapse with
#' `simplify(g, edge.attr.comb = list(weight = "sum"))`.
load_transit <- function(nodes = load_nodes(), edges = load_edges()) {
  g <- graph_from_data_frame(d = edges, directed = FALSE, vertices = nodes)
  igraph::E(g)$weight <- igraph::E(g)$capacity
  g
}

# 1. Demo ####################################################################
if (sys.nframe() == 0) {
  cat("\n\U0001F687 transit-multimodal (R)\n")
  cat("   Undirected multimodal transit; neighborhoods linked by metro & bus.\n\n")

  nodes <- load_nodes()
  edges <- load_edges()
  g <- load_transit(nodes, edges)

  cat(sprintf("✅ Loaded %d neighborhoods and %d edges (%d metro + %d bus).\n",
              nrow(nodes), nrow(edges),
              sum(edges$mode == "metro"), sum(edges$mode == "bus")))
  cat(sprintf("\U0001F517 Directed: %s | metro-served neighborhoods: %d | total seat capacity: %s riders/hr\n",
              is_directed(g), sum(nodes$has_metro),
              format(sum(edges$capacity), big.mark = ",")))
  cat("\U0001F389 Graph ready. Object `g` is an undirected, weighted, multimodal igraph.\n")
}
